### PR TITLE
Switch to CMake Build

### PR DESCRIPTION
**Pull request by jgoppert**
This is still a work in progress but it currently builds nuttx and runs the px4_simple_app. Posix and qurt support are in progress.

### CODE REVIEW

**Comment by jgoppert** [comment]
### Building

The cmake build is automated by a top level make file with convenience targets. The targets mimic the old make build system:

```bash
make px4fmu-v2_simple                                (builds firmware)
make px4fmu-v2_simple-upload                    (uploads firmware to board)
```

If you want to see all of the targets, or use the verbose flag, the best way is to go to the build folder created automatically, above.

```bash
cd build_px4fmu-v2_default
VERBOSE=1 make
make help                          (list all commands)
make test                           (do unit testing)
```

**Comment by LorenzMeier** [comment]
Fail:
```
cp: /Users/lm/src/Firmware/NuttX/nuttx/arch/arm/include/board: No such file or directory
cp: /Users/lm/src/Firmware/NuttX/nuttx/arch/arm/src/board: No such file or directory
cp: /Users/lm/src/Firmware/NuttX/nuttx/include/arch/board: No such file or directory
make[3]: *** [nuttx_copy_px4fmu-v2.stamp] Error 1
make[2]: *** [CMakeFiles/__nuttx_copy_px4fmu-v2.dir/all] Error 2
make[1]: *** [all] Error 2
make: *** [px4fmu-v2_simple] Error 2
```

**Comment by LorenzMeier** [comment]
After another git clean:
```bash
[ 92%] Generating romfs.bin
Pruning ROMFS files.
[100%] Generating romfs.o
romfs.o.c:2:20: error: redefinition of 'romfs_img_len'
 const unsigned int romfs_img_len = 0x;
                    ^
compilation terminated due to -Wfatal-errors.
make[3]: *** [src/romfs.o] Error 1
make[2]: *** [src/CMakeFiles/main.dir/all] Error 2
make[1]: *** [all] Error 2
make: *** [px4fmu-v2_simple] Error 2
```

**Comment by jgoppert** [comment]
@LorenzMeier, @mcharleb 

I will have to take a break for a few hours, so I'm going to write down what Lorenz and I just discussed in hopes that someone else can submit a fix while I'm away.

It looks like on darwin, the use of awk is causing an issue in the chunk of code that transforms the romfs binary image into an object file:

```cmake
	add_custom_command(OUTPUT ${OBJ}
		COMMAND ${ECHO} > ${OBJ}.c
		COMMAND ${CMAKE_C_COMPILER} ${CMAKE_C_FLAGS} -c ${OBJ}.c -o ${OBJ}.c.o
		COMMAND ${LD} -r -o ${OBJ}.bin.o ${OBJ}.c.o -b binary ${BIN}
		COMMAND ${NM} -p --radix=x ${OBJ}.bin.o
			| ${GREP} ${sym}_size
			| ${GREP} -o ^[0-9a-fA-F]*
			| ${AWK} "{print \"const unsigned int ${VAR}_len = 0x\"$1\";\"}" > ${OBJ}.c
		COMMAND ${CMAKE_C_COMPILER} ${CMAKE_C_FLAGS} -c ${OBJ}.c -o ${OBJ}.c.o
		COMMAND ${LD} -r -o ${OBJ} ${OBJ}.c.o ${OBJ}.bin.o
		COMMAND ${OBJCOPY} ${OBJ}
			--redefine-sym ${sym}_start=${VAR}
			--strip-symbol ${sym}_size
			--strip-symbol ${sym}_end
			--rename-section .data=.rodata
		# useful to comment remove statement when debugging
		COMMAND ${RM} ${OBJ}.c ${OBJ}.c.o ${OBJ}.bin.o
		DEPENDS ${BIN}
		VERBATIM
		)
```

This was translated from the original makefile here:
```make
define BIN_TO_OBJ
	@$(ECHO) "OBJ:     $2"
	@$(MKDIR) -p $(dir $2)
	$(Q) $(ECHO) > $2.c
	$(call COMPILE,$2.c,$2.c.o)
	$(Q) $(LD) -r -o $2.bin.o $2.c.o -b binary $1
	$(Q) $(ECHO) "const unsigned int $3_len = 0x`$(NM) -p --radix=x $2.bin.o | $(GREP) $(call BIN_SYM_PREFIX,$1)_size$$ | $(GREP) -o ^[0-9a-fA-F]*`;" > $2.c
	$(call COMPILE,$2.c,$2.c.o)
	$(Q) $(LD) -r -o $2 $2.c.o $2.bin.o
	$(Q) $(OBJCOPY) $2 \
		--redefine-sym $(call BIN_SYM_PREFIX,$1)_start=$3 \
		--strip-symbol $(call BIN_SYM_PREFIX,$1)_size \
		--strip-symbol $(call BIN_SYM_PREFIX,$1)_end \
		--rename-section .data=.rodata
	$(Q) $(REMOVE) $2.c $2.c.o $2.bin.o
endef
```

Note that awk is not used in the original makefile, which used inline bash evaluation. With cmake add_custom_command, it is not easy to save a variable or do inline evaluation, so I instead piped it into the expression using awk.


The build seems fine on linux machines, including the travis build: https://travis-ci.org/PX4/Firmware/builds/79253425

So it definetly looks operating system dependent, which is likely due to changes in gawk vs. awk on darwin. So it seems the best path forward is rewriting the bin_to_obj function in python, this will require pass tools(nm, grep, c compiler), and flags( c flags), to the python script , and then executing using subprocess.

File: cmake/builtin_commands.c.cmake [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/* builtin command list - automatically generated, do not edit */
```
Comment (vooon) [inline cmake/builtin_commands.c.cmake]: Why file extension is `.cmake`? Usually i seen `.in` for templates.

File: cmake/cmake_lists.jinja [hunk 2]
```diff
@@ -0,0 +1,73 @@
+############################################################################
```
Comment (vooon) [inline cmake/cmake_lists.jinja]: Note: project already uses EmPy (in `genmsg`), maybe better use that template engine?

File: cmake/px4_base.cmake [hunk 3]
```diff
@@ -0,0 +1,613 @@
...
+#		OUT			: the firmware target
+#
+#	Example:
+#		px4_add_upload(OUT upload
+#			BUNDLE main.px4
+#			)
+#
+function(px4_add_upload)
+	px4_parse_function_args(
+		NAME px4_generate_messages
```
Comment (vooon) [inline cmake/px4_base.cmake]: Wrong `NAME` arg.

File: cmake/px4_impl_nuttx.cmake [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+		endforeach()
+		if (MAIN)
+			set(builtin_apps_string
+				"${builtin_apps_string}\t{\"${MAIN}\", ${PRIORITY}, ${STACK}, ${MAIN}_main},\n")
+			set(builtin_apps_decl_string
+				"${builtin_apps_decl_string}extern int ${MAIN}_main(int argc, char *argv[]);\n")
+			math(EXPR command_count "${command_count}+1")
+		endif()
+	endforeach()
+	configure_file(${CMAKE_SOURCE_DIR}/cmake/builtin_commands.c.cmake
```
Comment (vooon) [inline cmake/px4_impl_nuttx.cmake]: Usually templates ends with `.in`.

**Comment by vooon** [comment]
I tested (on Ubuntu 15.04): it builds, but i think px4fmu-v2-simple is incomplete:

1. `builtin_commands.c`: contains only `uorb` and `px4_simple_app`.
2. `rc.autostart` are empty
3. ROMFS don't have a firmware for px4io

Also automatic fix of submodules don't work.

File: cmake/cmake_lists.jinja (hunk 2 above)
Comment (jgoppert) [inline cmake/cmake_lists.jinja]: This file is just used to drive cmake/convert_modules_to_cmake.py (convert module.mk's to CMakeLists.txt) so jinja is not used by the build process. So convert_modules_to_cmake.py will probably be deleted when we are fully done porting.

File: cmake/builtin_commands.c.cmake (hunk 1 above)
Comment (jgoppert) [inline cmake/builtin_commands.c.cmake]: Thanks, fixed.

File: cmake/px4_base.cmake (hunk 3 above)
Comment (jgoppert) [inline cmake/px4_base.cmake]: Thanks, fixed.

**Comment by jgoppert** [comment]
@vooon 

1. Those are the only apps added currently that have a command (see the module_list variable for simple config).
2. rc.autostart for me as well just contained commented text, so this may be an issue
3. Adding the px4io firmware is on the TODO list.

**Comment by jgoppert** [comment]
@voon, patches welcome.

**Comment by jgoppert** [comment]
@vooon patches welcome.

**Comment by jgoppert** [comment]
@LorenzMeier Can you try the nuttx build out on darwin now? I replaced the awk with a python script.

**Comment by jgoppert** [comment]
@mcharleb We have a link error on travis for the posix sitl, haven't been able to reproduce on my machine.
```bash
[35m[1mScanning dependencies of target lib__geo_lookup
[0m[100%] [32mBuilding C object src/lib/geo_lookup/CMakeFiles/lib__geo_lookup.dir/geo_mag_declination.c.o
[0m[31m[1mLinking C static library liblib__geo_lookup.a
[0m[100%] Built target lib__geo_lookup
[35m[1mScanning dependencies of target mainapp
[0m[100%] [32mBuilding CXX object src/firmware/posix/CMakeFiles/mainapp.dir/__/__/platforms/posix/main.cpp.o
[0m[31m[1mLinking CXX executable mainapp
[0m../../modules/mavlink/libmodules__mavlink.a(mavlink_messages.cpp.o): In function `MavlinkStreamSystemTime::send(unsigned long)':
mavlink_messages.cpp:(.text._ZN23MavlinkStreamSystemTime4sendEm[_ZN23MavlinkStreamSystemTime4sendEm]+0x10): undefined reference to `clock_gettime'
../../modules/mavlink/libmodules__mavlink.a(mavlink_messages.cpp.o): In function `_ZN23MavlinkStreamStatustext4sendEm.part.2':
mavlink_messages.cpp:(.text._ZN23MavlinkStreamStatustext4sendEm.part.2+0x136): undefined reference to `clock_gettime'
../../platforms/posix/px4_layer/libplatforms__posix__px4_layer.a(drv_hrt.c.o): In function `hrt_absolute_time':
drv_hrt.c:(.text.hrt_absolute_time+0x17): undefined reference to `clock_gettime'
drv_hrt.c:(.text.hrt_absolute_time+0x33): undefined reference to `clock_gettime'
../../modules/mavlink/libmodules__mavlink.a(mavlink_receiver.cpp.o): In function `MavlinkReceiver::handle_message_system_time(__mavlink_message*)':
mavlink_receiver.cpp:(.text._ZN15MavlinkReceiver26handle_message_system_timeEP17__mavlink_message+0x1b): undefined reference to `clock_gettime'
mavlink_receiver.cpp:(.text._ZN15MavlinkReceiver26handle_message_system_timeEP17__mavlink_message+0x68): undefined reference to `clock_settime'
../../modules/sdlog2/libmodules__sdlog2.a(sdlog2.c.o): In function `get_log_time_utc_tt':
sdlog2.c:(.text.get_log_time_utc_tt+0x16): undefined reference to `clock_gettime'
collect2: error: ld returned 1 exit status
make[3]: *** [src/firmware/posix/mainapp] Error 1
make[2]: *** [src/firmware/posix/CMakeFiles/mainapp.dir/all] Error 2
make[1]: *** [all] Error 2
make: *** [posix_sitl_simple] Error 2
travis_time:end:00d9184a:start=1441867515644944373,finish=1441867531039765654,duration=15394821281
[0K
[31;1mThe command "make posix_sitl_simple" exited with 2.[0m
travis_time:start:12ef1fe2
```

**Comment by mcharleb** [comment]
I'll have a look.

**Comment by jgoppert** [comment]
@LorenzMeier, @mcharleb nuttx firmware now building with cmake with default modules list except for uavcan. I can even connect over usb and also over qgroundcontrol. The only issue is that the parameters do not appear. I'm guessing we have an issue with romfs still?

**Comment by jgoppert** [comment]
@vooon The romfs generation was a bit mysterious to me so I tried not to make any significant changes, do you understand this enough to help rewrite it? It appears we have an issue with the current romfs as is and I'm not sure if this is related, but it might be worth trying.

**Comment by LorenzMeier** [comment]
The parameters will not load if anything with the system boot is wrong (and a recent version of Firmware will tell you in QGC). So essentially your boot aborted prematurely.

**Comment by vooon** [comment]
From ld black magick i see that equivalent:

```c
const uint8_t romfs[] = {
// content of romfs.img (genromfs)
};
const size_t romfs_len = sizeof romfs;
```

Perhaps you may simply generate romfs filler then compile to object file (see `bintoc`).
But linker magick may be faster...

**Comment by vooon** [comment]
Example: `xxd -i romfs.img`

But i'm not sure how portable that program. Does it exist on apple?

**Comment by jgoppert** [comment]
@LorenzMeier, @mcharleb , @vooon  The build is very smooth now. Mark and I really spent some time to cleanup and simplify the confiig files. The only issues left are:

### Nuttx
* Params don't load. This causes failed boot when autostart is enabled. When the autostart script is empty, I can communicate over usb serial with nsh.
* Have to build/attach px4io firmware.
* Need to fix uavcan module building

### Posix SITL
* Segfaults on some simulated sensors (not sure if this was an existing behaviour or not).
* Would be nice to get param handling out of linker so it would build on mac.

### Build Update

```bash
make px4fmu-v2_default                                 (build)
make px4fmu-v2_default upload                          (upload)
make px4fmu-v2_default test                            (test)
make px4fmu-v2_default package                         (create zip of binary)
make px4fmu-v2_default package_source                  (create zip of source)

same for other targets:
make posix_sitl_simple upload
etc.
```

**Comment by jgoppert** [comment]
Success, travis build passed.

**Comment by vooon** [comment]
@jgoppert where? :)

`px4fmu-v2_default` successfully builds, but 'package' target generate empty `px4-1.0.1-Generic.zip`.
Also where i can find binary for flasher?

**Comment by jgoppert** [comment]
@vooon I hadn't added any install targets yet, I just added the firmware as an install target for nuttx, so it should work now. To upload you can use make target upload. Alternatively it is in: /build_px4fmu-v2_default/src/firmware/nuttx/nuttx_px4fmu-v2_default.px4

**Comment by LorenzMeier** [comment]
Closing for upstream branch.

**Pull request closed without merging**